arbitrator/wasm-libraries/go-js/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+        let source = self.values.id_to_value(source);
+        let new_value = self.values.id_to_value(new_value);
+        match &source {
+            JsValue::Array(array) => {
+                let mut array = array.lock();
+                if index >= array.len() {
+                    array.resize(index + 1, JsValue::Undefined);
+                }
+                array[index] = new_value;
+            }
+            JsValue::Uint8Array(array) => {
+                let mut array = array.lock();
+                let new_value = match new_value {
+                    JsValue::Number(x) => x as u8,
+                    _ => {
+                        eprintln!("Go is setting a Uint8Array index to {new_value:?}");
+                        0
+                    }
+                };
+                if index >= array.len() {
+                    eprintln!("Go is setting out-of-range index {index} in Uint8Array of size {} to {new_value:?}", array.len());
+                } else {
+                    array[index] = new_value;
+                }
+            }
+            _ => {
+                panic!("Go attempted to call valueSetIndex on invalid type: {source:?}");
+            }
+        }
+    }
+
+    pub fn value_call<'a>(
+        &self,
+        env: &'a mut (dyn JsEnv + 'a),
+        object: JsValueId,
+        method: &str,
+        args: &[JsValueId],
+    ) -> eyre::Result<JsValueId> {
+        let this = self.values.id_to_value(object);
+        let object = this.clone().assume_object("valueCall target");
+        let JsValue::Function(function) = object.get(method) else {
+            panic!("Go attempted to call {object:?} non-function field {method}");
+        };
+        let args = args.iter().map(|x| self.values.id_to_value(*x)).collect();
+        let result = function.call(env, this, args)?;
+        Ok(self.values.assign_id(result))
+    }
+
+    pub fn value_new<'a>(
+        &self,
+        env: &'a mut (dyn JsEnv + 'a),
+        constructor: JsValueId,
+        args: &[JsValueId],
+    ) -> eyre::Result<JsValueId> {
+        // All of our constructors are normal functions that work via a call
+        let JsValue::Function(function) = self.values.id_to_value(constructor) else {
+            panic!("Go attempted to construct non-function {constructor:?}");
+        };
+        let args = args.iter().map(|x| self.values.id_to_value(*x)).collect();
+        let result = function.call(env, JsValue::Undefined, args)?;
+        Ok(self.values.assign_id(result))
+    }
+
+    pub fn string_val(&self, s: String) -> JsValueId {
+        self.values.assign_id(JsValue::String(Arc::new(s)))
+    }
+
+    pub fn value_length(&self, array: JsValueId) -> usize {
+        let len = match self.values.id_to_value(array) {
+            JsValue::Array(array) => array.lock().len(),
+            JsValue::Uint8Array(array) => array.lock().len(),
+            x => {
```

Comment:
Go expects an integer rather than a type, so we should probs panic in our case if the result is not a number. I've updated the panic message to clarify this a bit, as well as added an implementation for `String`.